System fonts are not displayed in the correct style for their idiom in compatibility mode on visionOS.

megangardner opened the pull request:
#### 7ef05e80175b3e2596090dedec3c93e5ea306796
<pre>
System fonts are not displayed in the correct style for their idiom in compatibility mode on visionOS.
<a href="https://bugs.webkit.org/show_bug.cgi?id=259844">https://bugs.webkit.org/show_bug.cgi?id=259844</a>
rdar://107389870

Reviewed by Tim Horton.

When requesting information about system fonts, we were always going with the current idiom, which for
the web-process will always match the device. If the UIProcess is in compatibility mode, we need
to make sure that we are using the idiom that it is in, or we will get system fonts on the wrong size.

We move the required classes into PAL so they can be used from WebCore while maintaining platform agnosticism
in WebCore itself, and then read the appropriate flag in the font code. Since the font database is
multithreaded, we also need to ensure that the flag is thread-safe.

* Source/WebCore/PAL/pal/system/ios/Device.cpp: Renamed from Source/WebCore/platform/ios/Device.cpp.
(PAL::deviceClassIsSmallScreen):
(PAL::deviceClassIsReality):
(PAL::deviceName):
(PAL::deviceHasIPadCapability):
* Source/WebCore/PAL/pal/system/ios/Device.h: Renamed from Source/WebCore/platform/ios/Device.h.
* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.h: Renamed from Source/WebKit/Shared/UserInterfaceIdiom.h.
* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.mm: Renamed from Source/WebKit/Shared/UserInterfaceIdiom.mm.
(PAL::currentUserInterfaceIdiomIsSmallScreen):
(PAL::currentUserInterfaceIdiomIsReality):
(PAL::currentUserInterfaceIdiom):
(PAL::setCurrentUserInterfaceIdiom):
(PAL::updateCurrentUserInterfaceIdiom):

Canonical link: <a href="https://commits.webkit.org/266710@main">https://commits.webkit.org/266710@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe32bd8727739a3cb2bce5ce59c02f4e58b2be49

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14537 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14848 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15192 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16284 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/13745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14673 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17363 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14925 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14719 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15242 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12346 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17013 "Built successfully") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/14641 "webkitpy-tests (failure)") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12525 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13111 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/20116 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13604 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13276 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16508 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/13825 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13828 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/11653 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/14582 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13119 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/12968 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3771 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3515 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17457 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/14971 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13672 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3587 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.cpp:
```diff
@@ -392,7 +393,7 @@ auto SystemFontDatabase::platformSystemFontShorthandInfo(FontShorthand fontShort
     auto interrogateTextStyleShorthandItem = [] (CFStringRef textStyle) {
         CGFloat weight = 0;
-        float size = CTFontDescriptorGetTextStyleSize(textStyle, contentSizeCategory(), kCTFontTextStylePlatformDefault, &weight, nullptr);
+        float size = CTFontDescriptorGetTextStyleSize(textStyle, contentSizeCategory(), PAL::currentUserInterfaceIdiomIsReality() ? kCTFontTextStylePlatformDefault : kCTFontTextStylePlatformPhone, &weight, nullptr);
```
This seems wrong: on *all* non-visionOS platforms you're going to say kCTFontTextStylePlatformPhone

hortont424 replied:
And with the minor complexity that fixing that is going to add, can you maybe shuffle it out to a static function outside of this one?

aprotyas commented on Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.mm:
```diff
@@ -84,7 +86,7 @@ bool updateCurrentUserInterfaceIdiom()
         }
 
-        return UserInterfaceIdiom::Default;
+        return PAL::UserInterfaceIdiom::Default;
```
Nit: I think you don't have to reference the `PAL` namespace, since you're already in it.

hortont424 commented on Source/WebGPU/WebGPU/Adapter.mm:
```diff
@@ -32,6 +32,7 @@
 #import <algorithm>
 #import <wtf/StdLibExtras.h>
 
+
```
?

aprotyas replied:
Nit: can remove this newline

hortont424 commented on Source/WebCore/platform/graphics/SystemFontDatabase.cpp:
```diff
@@ -33,7 +33,6 @@ SystemFontDatabase::SystemFontDatabase() = default;
 auto SystemFontDatabase::systemFontShorthandInfo(FontShorthand fontShorthand) -> const SystemFontShorthandInfo& {
     if (auto& entry = m_systemFontShorthandCache[fontShorthand])
         return *entry;
-
```
?

aprotyas reviewed (commented):
Not a real review, just a couple of optional comments.

hortont424 commented on Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.mm:
```diff
@@ -28,32 +28,34 @@
-static std::optional<UserInterfaceIdiom> s_currentUserInterfaceIdiom;
+namespace PAL {
+
+static std::atomic<std::optional<UserInterfaceIdiom>> s_currentUserInterfaceIdiom;
```
Would be great to check with folks that this does what we hope it does.

megangardner replied:
Abrar and I had a thread about WTF::Atomic, and it's just dealing with a case with swapping that isn't applicable here.

mwyrzykowski replied:
I am not so familiar with this code but instead of using `atomic<optional<UserInterfaceIdiom>>` couldn't you use `atomic<UserInterfaceIdiom>` and default `s_currentUserInterfaceIdiom = Unknown` (where Unknown is a new value in the enum) and then instead of checking `!s_.load()` check `s_.load() != Unknown`? Or if we don't want to expand `UserInterfaceIdiom` then we could use `atomic<someIntegerType>` and perform conversions?

megangardner replied:
We've talked with the CPP folks, and this usage is fine.

hortont424 commented on Source/WebCore/platform/ios/PlatformScreenIOS.mm:
```diff
@@ -39,10 +38,13 @@
 #import <pal/spi/ios/UIKitSPI.h>
+#import <pal/system/ios/Device.h>
+
+#import <pal/cocoa/MediaToolboxSoftLink.h>
```
The old layout was more correct (smooshed against the bottom of the quoted includes, in the same paragraph, sorted as `sort` would.

hortont424 replied:
(except for the soft links!)

megangardner replied:
The bottom is a SoftLink, so I'm not sure what you're saying here.

megangardner committed:
System fonts are not displayed in the correct style for their idiom in compatibility mode on visionOS.
https://bugs.webkit.org/show_bug.cgi?id=259844
rdar://107389870

Reviewed by Tim Horton.

When requesting information about system fonts, we were always going with the current idiom, which for
the web-process will always match the device. If the UIProcess is in compatibility mode, we need
to make sure that we are using the idiom that it is in, or we will get system fonts on the wrong size.

We move the required classes into PAL so they can be used from WebCore while maintaining platform agnosticism
in WebCore itself, and then read the appropriate flag in the font code. Since the font database is
multithreaded, we also need to ensure that the flag is thread-safe.

* Source/WebCore/PAL/pal/system/ios/Device.cpp: Renamed from Source/WebCore/platform/ios/Device.cpp.
(PAL::deviceClassIsSmallScreen):
(PAL::deviceClassIsReality):
(PAL::deviceName):
(PAL::deviceHasIPadCapability):
* Source/WebCore/PAL/pal/system/ios/Device.h: Renamed from Source/WebCore/platform/ios/Device.h.
* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.h: Renamed from Source/WebKit/Shared/UserInterfaceIdiom.h.
* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.mm: Renamed from Source/WebKit/Shared/UserInterfaceIdiom.mm.
(PAL::currentUserInterfaceIdiomIsSmallScreen):
(PAL::currentUserInterfaceIdiomIsReality):
(PAL::currentUserInterfaceIdiom):
(PAL::setCurrentUserInterfaceIdiom):
(PAL::updateCurrentUserInterfaceIdiom):

Canonical link: https://commits.webkit.org/266710@main
